lib/pos_mempool_ledger.go

```diff
@@ -0,0 +1,101 @@
+package lib
+
+import (
+	"github.com/pkg/errors"
+	"math"
+	"sync"
+)
+
+// BalanceLedger is a simple in-memory ledger of reserved balances for user public keys. The values in the ledger can be
+// increased or decreased, as long as user's reserved balance doesn't exceed the user's spendable balance.
+type BalanceLedger struct {
+	sync.RWMutex
+
+	// Map of public keys to reserved balances in nanos.
+	reservedBalancesNanos map[PublicKey]uint64
```

Comment:
Cool idea! Abstracting away a thread-safe map into a reusable struct is a clean solution. However, as I've started implementing this generic and using it in BalanceLedger, I've encountered some issues. The main issue occurred when I tried rewriting the `IncreaseBalance` function. In this case, we need to perform a read and then a write. We first read the user's balance and then write the balance + amount. The problem is that the `RwLockMap` functions `Get` and `Set` release locks. So, theoretically, some undesirable instruction could sneak its way in between and 180 the balance. We would need something that does a Get and a Set under a single lock.

To solve this, I came up with an `Update` method:
```
func (m *RWLockMap[K, V]) Update(key K, expression func(V, bool) V) {
	m.Lock()
	defer m.Unlock()

	value, exists := m._innerMap[key]
	newValue := expression(value, exists)

	m._innerMap[key] = newValue
}
```
This method works nicely because we can encapsulate the "+ amount" logic inside the `expression` argument. But even with this function, some things become problematic. Let's say we want to check for the overflow after a read but before the write. In such a case, we want to return an error. However, this means we would need to do this inside the `expression`, which should, in turn, actually return `(V, error)`. Similarly, `Update` should now return an `error`. More complexity. 

Though I'm chill with complexity. What do you think? If you're fine with having the `Update` method, I'm happy to add this generic.

Reply:
In this case, migrating to a generic RWLockMap data structure may be overkill. You make the call on whether to implement it or not @AeonSw4n. IMO, the code reads clearly as-is and the amount of code de-duplication we gain by implementing the RWLockMap is minimal.